File: Sample/SDOSHeroExample/App/Controllers/Controllers/Example/SDOSHeroDetailViewController.m
//
//  SDOSHeroDetailViewController.m
//  SDOSHeroExample
//
//  Created by Antonio Jesús Pallares on 13/11/17.
//  Copyright © 2017 SDOS. All rights reserved.
//

#import "SDOSHeroDetailViewController.h"
#import "SDOSHeroExample-Swift.h"


@interface SDOSHeroDetailViewController ()
@property (weak, nonatomic) IBOutlet UILabel *lbAnimationGoTitle;
@property (weak, nonatomic) IBOutlet UILabel *lbAnimationGo;
@property (weak, nonatomic) IBOutlet UILabel *lbAnimationReturnTitle;
@property (weak, nonatomic) IBOutlet UILabel *lbAnimationReturn;

@property (weak, nonatomic) IBOutlet NSLayoutConstraint *constraintCenterTop;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *constraintCenterBottom;

@end

@implementation SDOSHeroDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self loadStyle];
}


- (void)loadStyle {
    if (self.navigationController.viewControllers.firstObject == self) {
        self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(actionBack:)];
    }
    UIBarButtonItem *item0 = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(actionNext:)];
    UIBarButtonItem *item1 = [[UIBarButtonItem alloc] initWithTitle:@"Root" style:UIBarButtonItemStylePlain target:self action:@selector(actionGoToRootViewController:)];
    self.navigationItem.rightBarButtonItems = @[item0, item1];
    
    
    [self.lbAnimationGo loadStyleLabelExample];
    self.lbAnimationGo.sdosHeroID = self.textAnimationType;
    [self.lbAnimationGo setSDOSHeroModifiers:@[HeroModifierSpringWith(300, 15)]];
    self.lbAnimationGo.text = self.textAnimationType;

    NSString *textOppositeAnimationType = [SDOSHeroAnimationTypeUtil identifierForType:SDOSHeroOppositeAnimationTo([SDOSHeroAnimationTypeUtil typeForIdentifier:self.textAnimationType])];
    
    
    [self.lbAnimationReturn loadStyleLabelExample];
    self.lbAnimationReturn.sdosHeroID = textOppositeAnimationType;
    [self.lbAnimationReturn setSDOSHeroModifiers:@[HeroModifierDelay(0.5)]];
    self.lbAnimationReturn.text = textOppositeAnimationType;
    
    NSString *titleGo;
    NSString *titleReturn;
    if ([self isInsideNavigation]) {
        titleGo = @"The push animation was";
        titleReturn = @"The pop animation will be";
    } else {
        titleGo = @"The present animation was";
        titleReturn = @"The dismiss animation will be";
    }
    
    [self.lbAnimationGoTitle loadStyleLabelExample];
    self.lbAnimationGoTitle.text = titleGo;
    
    [self.lbAnimationReturnTitle loadStyleLabelExample];
    self.lbAnimationReturnTitle.text = titleReturn;
    
    
    if (arc4random() % 2) {
        self.constraintCenterTop.priority = 1;
        self.constraintCenterBottom.priority = 999;
    }
}


#pragma mark - Actions

- (void)actionBack:(UIBarButtonItem *)sender {
    if (self.navigationController.viewControllers.firstObject == self) {
        [self.navigationController dismissViewControllerAnimated:YES completion:nil];
    } else {
        [self.navigationController popViewControllerAnimated:YES];
    }
}


- (void)actionNext:(UIBarButtonItem *)sender {
    SDOSHeroAnimationType type = [SDOSHeroAnimationTypeUtil randomAnimation];
    if (self.navigationController.viewControllers.firstObject == self) {
        UIViewController *vc = [self.storyboard instantiateViewControllerWithIdentifier:SDOSHeroDetailNavigationViewControllerIdentifier];
        
        vc.sdosHeroIsEnabled = YES;
        [vc setSDOSHeroAnimationTypeForModalPresenting:type];
        if ([vc isKindOfClass:[UINavigationController class]]) {
            UIViewController *firstVC = ((UINavigationController *)vc).viewControllers.firstObject;
            if ([firstVC isKindOfClass:[SDOSHeroDetailViewController class]]) {
                ((SDOSHeroDetailViewController *) firstVC).textAnimationType = [SDOSHeroAnimationTypeUtil identifierForType:type];
            }
        }
        
        [self.navigationController presentViewController:vc animated:YES completion:nil];
    } else {
        UIViewController *vc = [self.storyboard instantiateViewControllerWithIdentifier:SDOSHeroDetailViewControllerIdentifier];
        if ([vc isKindOfClass:[SDOSHeroDetailViewController class]]) {
            ((SDOSHeroDetailViewController *) vc).textAnimationType = [SDOSHeroAnimationTypeUtil identifierForType:type];
        }
        [(SDOSHeroNavigationController *)self.navigationController pushViewController:vc usingAnimation:type];
    }
}


- (void)actionGoToRootViewController:(UIBarButtonItem *)sender {
    if (self.navigationController.viewControllers.firstObject == self) {
        [self sdoshero_unwindToRootViewController];
    } else {
        [self.navigationController popToRootViewControllerAnimated:YES];
    }
}


#pragma mark - Helper

- (BOOL)isInsideNavigation {
    return (self.navigationController.viewControllers.firstObject != self);
}

@end

File: Sample/SDOSHero/App/Controllers/Controllers/Example/SDOSHeroDetailViewController.swift
//
//  SDOSHeroDetailViewController.swift
//  SDOSHeroExample
//
//  Created by Antonio Jesús Pallares on 02/05/2019.
//  Copyright © 2019 SDOS. All rights reserved.
//

import Foundation
import Hero
import SDOSHero

class SDOSHeroDetailViewController: BaseViewController {
    
    var textAnimationType = ""
    var heroIdPrefix = ""
    
    @IBOutlet private weak var lbAnimationGoTitle: UILabel!
    @IBOutlet private weak var lbAnimationGo: UILabel!
    @IBOutlet private weak var lbAnimationReturnTitle: UILabel!
    @IBOutlet private weak var lbAnimationReturn: UILabel!
    
    @IBOutlet private weak var constraintCenterTop: NSLayoutConstraint!
    @IBOutlet private weak var constraintCenterBottom: NSLayoutConstraint!
    
    private var isInsideNavigation: Bool {
        return navigationController?.viewControllers.first !== self
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        loadStyle()
    }
    
    private func loadStyle() {
        if navigationController?.viewControllers.first === self {
            navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(actionBack(_ :)))
        }
        if #available(iOS 11.0, *) {
            navigationItem.largeTitleDisplayMode = .never
        }
        
        let item0 = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(actionNext(_ :)))
        let item1 = UIBarButtonItem(title: "Root", style: .plain, target: self, action: #selector(actionGoToRootViewController(_ :)))
        navigationItem.rightBarButtonItems = [item0, item1]
        
        lbAnimationGo.loadStyleLabelExample()
        lbAnimationGo.hero.id = heroIdPrefix + textAnimationType
        lbAnimationGo.hero.modifiers = [.spring(stiffness: 300, damping: 15)]
        lbAnimationGo.text = textAnimationType
        
        let animationType = HeroDefaultAnimationType.typeForIdentifier(textAnimationType)
        let textOppositeAnimationType = animationType.oppositeAnimationType.identifier
        lbAnimationReturn.loadStyleLabelExample()
        lbAnimationReturn.hero.id = heroIdPrefix + textOppositeAnimationType
        lbAnimationReturn.hero.modifiers = [.delay(0.5)]
        lbAnimationReturn.text = textOppositeAnimationType
        
        let titleGo: String
        let titleReturn: String
        if isInsideNavigation {
            titleGo = "The push animation was"
            titleReturn = "The pop animation will be"
        } else {
            titleGo = "The present animation was";
            titleReturn = "The dismiss animation will be";
        }
        
        lbAnimationGoTitle.loadStyleLabelExample()
        lbAnimationGoTitle.text = titleGo
        
        lbAnimationReturnTitle.loadStyleLabelExample()
        lbAnimationReturnTitle.text = titleReturn
        
        if Bool.random() {
            constraintCenterTop.priority = UILayoutPriority(1)
            constraintCenterBottom.priority = UILayoutPriority(999)
        }
    }
    
    @objc func actionBack(_ sender: Any) {
        if isInsideNavigation {
            navigationController?.popViewController(animated: true)
        } else {
            navigationController?.dismiss(animated: true, completion: nil)
        }
    }
    
    @objc func actionNext(_ sender: Any) {
        let animationType = HeroDefaultAnimationType.allPosibleAnimationTypes.randomElement() ?? .auto
        if isInsideNavigation {
            if let detailVC = storyboard?.instantiateViewController(withIdentifier: SDOSHeroDetailViewControllerIdentifier) as? SDOSHeroDetailViewController {
                detailVC.textAnimationType = animationType.identifier
                (navigationController as? SDOSHeroNavigationController)?.pushViewController(detailVC, usingAnimation: animationType)
            }
        } else {
            if
                let navigationController = storyboard?.instantiateViewController(withIdentifier: SDOSHeroDetailNavigationViewControllerIdentifier) as? UINavigationController,
                let firstVC = navigationController.viewControllers.first as? SDOSHeroDetailViewController
            {
                navigationController.hero.isEnabled = true
                navigationController.animationTypeForPresenting = animationType
                firstVC.textAnimationType = animationType.identifier
                navigationController.present(navigationController, animated: true, completion: nil)
            }
        }
    }
    
    @objc func actionGoToRootViewController(_ sender: Any) {
        if isInsideNavigation {
            navigationController?.popToRootViewController(animated: true)
        } else {
            dismissAllPresentedViewControllers()
        }
    }
    
}
